oidc-lib/src/main/java/nl/finalist/liferay/oidc/LibFilter.java

```diff
@@ -72,87 +71,66 @@ public LibFilter(LiferayAdapter liferay) {
-                        // From there, the request should be redirected back to the Liferay portal home page.
-                        request.getSession().invalidate();
-                        redirectToLogout(request, response, ssoLogoutUri, ssoLogoutParam, ssoLogoutValue);
-                        // no continuation of the filter chain; we expect the redirect to commence.
-                        return FilterResult.BREAK_CHAIN;
-                    }
+        if (pathInfo.contains("/portal/login")) {
+            if (!StringUtils.isBlank(request.getParameter(REQ_PARAM_CODE))
+                    && !StringUtils.isBlank(request.getParameter(REQ_PARAM_STATE))) {
+
+                if (!isUserLoggedIn(request)) {
+                    // LOGIN: Second time it will expect a code and state param to be set, and will exchange the code for an access token.
+                    liferay.trace("About to exchange code for access token");
+                    exchangeCodeForAccessToken(request);
+                } else {
+                    liferay.trace("subsequent run into filter during openid conversation, but already logged in." +
+                            "Will not exchange code for token twice.");
                 }
-		}
+            } else {
+                // LOGIN: The first time this filter gets hit, it will redirect to the OP.
+                liferay.trace("About to redirect to OpenID Provider");
+                redirectToLogin(request, response, oidcConfiguration.clientId());
+                // no continuation of the filter chain; we expect the redirect to commence.
+                return FilterResult.BREAK_CHAIN;
+            }
+        } else if (pathInfo.contains("/portal/logout")) {
+            final String ssoLogoutUri = oidcConfiguration.ssoLogoutUri();
+            final String ssoLogoutParam = oidcConfiguration.ssoLogoutParam();
+            final String ssoLogoutValue = oidcConfiguration.ssoLogoutValue();
+            if (null != ssoLogoutUri && ssoLogoutUri.length() > 0 && isUserLoggedIn(request)) {
+
+                liferay.trace("About to logout from SSO by redirect to " + ssoLogoutUri);
+                // LOGOUT: If Portal Logout URL is requested, redirect to OIDC Logout resource afterwards to globally logout.
+                // From there, the request should be redirected back to the Liferay portal home page.
+                request.getSession().invalidate();
+                redirectToLogout(request, response, ssoLogoutUri, ssoLogoutParam, ssoLogoutValue);
+                // no continuation of the filter chain; we expect the redirect to commence.
+                return FilterResult.BREAK_CHAIN;
+            }
+        }
+
         // continue chain
-		return FilterResult.CONTINUE_CHAIN;
+        return FilterResult.CONTINUE_CHAIN;
 
     }
 
-    private boolean isContainsSite(String serverName, List<String> list) {
```

Comment:
removed fonction since I use a Set as dataStructure that has this feature for free